### PR TITLE
Remove Open3 and move process forks

### PR DESCRIPTION
**Pull request by dhollinger**
Due to issues with the webhook script hitting Gitlab and Github webhook
timeout limits waiting for r10k or mco to complete, the Open3 calls
have been removed entirely from the run_command method. 

Additionally, to ensure that all commands/APIs run asynchronously, the Process.detach() has been
moved into the method calls in each POST endpoint.

### CODE REVIEW

File: templates/webhook.bin.erb [hunk 1]
```diff
@@ -117,7 +117,7 @@ class Server < Sinatra::Base
 
     module_name = sanitize_input(module_name)
     $logger.info("Deploying module #{module_name}")
-    deploy_module(module_name)
+    Process.detach(fork { deploy_module(module_name) })
```
[the remark this answers is not in the document]
Comment (dhollinger) [reply templates/webhook.bin.erb]: I removed the forking of the process inside the `run_command` method. Any potential API calls (like the MCO ruby client) also need to be forked or there will be issues with Gitlab/Github Webhook timeouts.

File: templates/webhook.bin.erb [hunk 2]
```diff
@@ -253,14 +253,9 @@ class Server < Sinatra::Base
         # the other one to complete.
         file.flock(File::LOCK_EX)
 
-        if Open3.respond_to?('capture3')
-          stdout, stderr, exit_status = Open3.capture3(command)
-          message = "triggered: #{command}\n#{stdout}\n#{stderr}"
```
[the remark this answers is not in the document]
Comment (dhollinger) [reply templates/webhook.bin.erb]: The problem with the `Open3` calls is that in any case where a user has aggregate webhooks, webhooks that trigger other webhooks, or API calls in an app that trigger a webhook that checks the response code, Gitlab/Github will always return an `HTTP 500` if the command took longer than 10 seconds to execute, which in turn, invalidates the HTTP code the webhook returns.

I've spent a few weeks trying to figure this out and have narrowed it down to `Open3` forcing the API call to hold on to the connection until the command is complete. 

I spent over an hour last night testing logging, etc with the changes and I continued to get logs in the main webhook log. We can make forking a non-default, but as of now, any user that has an r10k environment that takes more than 10 seconds to complete will get an HTTP 500 regardless of what the webhook sends back to Github/Gitlab and potentially Bitbucket (can't find docs on what the timeout is).

File: templates/webhook.bin.erb [hunk 3]
```diff
@@ -205,7 +205,7 @@ class Server < Sinatra::Base
       return 200
     else
       $logger.info("Deploying environment #{env}")
-      deploy(env, deleted)
+      Process.detach(fork { deploy(env, deleted) })
```
[the remark this answers is not in the document]
Comment (dhollinger) [reply templates/webhook.bin.erb]: I removed the forking of the process inside the run_command method. Any potential API calls (like the MCO ruby client) also need to be forked or there will be issues with Gitlab/Github Webhook timeouts.

File: templates/webhook.bin.erb [hunk 4]
```diff
@@ -253,14 +253,9 @@ class Server < Sinatra::Base
...
 
-        if Open3.respond_to?('capture3')
-          stdout, stderr, exit_status = Open3.capture3(command)
-          message = "triggered: #{command}\n#{stdout}\n#{stderr}"
-        else
-          message = "forked: #{command}"
-          Process.detach(fork{ exec "#{command} &"})
-          exit_status = 0
-        end
+        message = "forked: #{command}"
```
Comment (rnelson0) [inline templates/webhook.bin.erb]: I think I trusted this log message and assumed it was forked, missing the removal of the `Process.detach` portion! This should be updated to say `executed:` or similar.

File: templates/webhook.bin.erb (hunk 2 above)
[the remark this answers is not in the document]
Comment (rnelson0) [reply templates/webhook.bin.erb]: That is good to hear about the logging.

I believe an immediate failure would still return a 200 with this change. Is there another way to allow those that complete in <10s to return a 500 if the r10k command fails? If not, then an async check around the other fork sections would preserve that.

[the remark this answers is not in the document]
Comment (binford2k) [reply templates/webhook.bin.erb]: It's possible, though getting more complex. It would look something like this absolutely untested code that *might leave zombies behind!*

```ruby
require 'timeout'

begin  
 Timeout.timeout(10) do
  process = Process.detach(fork { deploy_module(module_name) })
  Process.wait process.pid
  message = "deployed with: #{command}"
 end
rescue Timeout::Error
  message = "forked deploy into background: #{command}"
end
```

File: templates/webhook.bin.erb (hunk 4 above)
Comment (dhollinger) [reply templates/webhook.bin.erb]: The process is still being forked and running in the background, but I can replace that with "executed" if you'd like

File: templates/webhook.bin.erb [hunk 5]
```diff
@@ -253,14 +253,9 @@ class Server < Sinatra::Base
...
 
-        if Open3.respond_to?('capture3')
-          stdout, stderr, exit_status = Open3.capture3(command)
-          message = "triggered: #{command}\n#{stdout}\n#{stderr}"
-        else
-          message = "forked: #{command}"
-          Process.detach(fork{ exec "#{command} &"})
-          exit_status = 0
-        end
+        message = "forked: #{command} in the background"
```
Comment (rnelson0) [inline templates/webhook.bin.erb]: `message = "triggered: #{command}"` would be better here IMO.

Comment (dhollinger) [reply templates/webhook.bin.erb]: Updated
